tlsutil/config.go

```diff
@@ -233,156 +234,186 @@ func (c *Configurator) commonTLSConfig(additionalVerifyIncomingFlag bool) (*tls.
+	config.GetConfigForClient = func(*tls.ClientHelloInfo) (*tls.Config, error) {
+		return c.IncomingRPCConfig(), nil
+	}
+	return config
 }
 
 // IncomingHTTPSConfig generates a *tls.Config for incoming HTTPS connections.
-func (c *Configurator) IncomingHTTPSConfig() (*tls.Config, error) {
-	return c.commonTLSConfig(c.base.VerifyIncomingHTTPS)
+func (c *Configurator) IncomingHTTPSConfig() *tls.Config {
+	c.log("IncomingHTTPSConfig")
+	c.RLock()
+	defer c.RUnlock()
+	config := c.commonTLSConfig(c.base.VerifyIncomingHTTPS)
+	config.GetConfigForClient = func(hello *tls.ClientHelloInfo) (*tls.Config, error) {
+		config := c.IncomingHTTPSConfig()
+		config.NextProtos = hello.SupportedProtos
+		return config, nil
+	}
+	return config
 }
 
 // IncomingTLSConfig generates a *tls.Config for outgoing TLS connections for
 // checks. This function is seperated because there is an extra flag to
 // consider for checks. EnableAgentTLSForChecks and InsecureSkipVerify has to
 // be checked for checks.
-func (c *Configurator) OutgoingTLSConfigForCheck(id string) (*tls.Config, error) {
+func (c *Configurator) OutgoingTLSConfigForCheck(skipVerify bool) *tls.Config {
+	c.log("OutgoingTLSConfigForCheck")
+	c.RLock()
+	defer c.RUnlock()
 	if !c.base.EnableAgentTLSForChecks {
 		return &tls.Config{
-			InsecureSkipVerify: c.getSkipVerifyForCheck(id),
-		}, nil
+			InsecureSkipVerify: skipVerify,
+		}
 	}
 
-	tlsConfig, err := c.commonTLSConfig(false)
-	if err != nil {
-		return nil, err
-	}
-	tlsConfig.InsecureSkipVerify = c.getSkipVerifyForCheck(id)
-	tlsConfig.ServerName = c.base.ServerName
-	if tlsConfig.ServerName == "" {
-		tlsConfig.ServerName = c.base.NodeName
+	config := c.commonTLSConfig(false)
+	config.InsecureSkipVerify = skipVerify
+	config.ServerName = c.base.ServerName
+	if config.ServerName == "" {
+		config.ServerName = c.base.NodeName
 	}
 
-	return tlsConfig, nil
+	return config
 }
 
 // OutgoingRPCConfig generates a *tls.Config for outgoing RPC connections. If
 // there is a CA or VerifyOutgoing is set, a *tls.Config will be provided,
 // otherwise we assume that no TLS should be used.
-func (c *Configurator) OutgoingRPCConfig() (*tls.Config, error) {
-	useTLS := c.base.CAFile != "" || c.base.CAPath != "" || c.base.VerifyOutgoing
-	if !useTLS {
-		return nil, nil
+func (c *Configurator) OutgoingRPCConfig() *tls.Config {
+	c.log("OutgoingRPCConfig")
+	c.RLock()
+	defer c.RUnlock()
+	if c.outgoingRPCTLSDisabled() {
+		return nil
 	}
 	return c.commonTLSConfig(false)
 }
 
+func (c *Configurator) outgoingRPCTLSDisabled() bool {
+	return c.base.CAFile == "" && c.base.CAPath == "" && !c.base.VerifyOutgoing
+}
+
 // OutgoingRPCWrapper wraps the result of OutgoingRPCConfig in a DCWrapper. It
 // decides if verify server hostname should be used.
 func (c *Configurator) OutgoingRPCWrapper() (DCWrapper, error) {
-	// Get the TLS config
-	tlsConfig, err := c.OutgoingRPCConfig()
-	if err != nil {
-		return nil, err
-	}
```

Comment:
And this is another important line because creating the actual configuration is postponed until later. Otherwise it outgoing connections would never pick up changes in the configuration.